### ISSUE TITLE
Document when to choose terraform-google-project-factory vs cloud-foundation-fabric

### ISSUE THREAD

**Issue by jawnsy** [terraform-google-modules/cloud-foundation-fabric#361]
### TL;DR

As a user trying to adopt Terraform best practices and secure our Google Cloud account, it's difficult to understand whether we should choose modules from terraform-google-project-factory or cloud-foundation-fabric. It would be nice to have cross-referencing documentation describing when a user should choose each approach.

### Additional information

[cloud-foundation-fabric v0.1](https://github.com/terraform-google-modules/cloud-foundation-fabric/releases/tag/v0.1) was published Mar 24, 2020. In contrast, [project-factory v0.1.0](https://github.com/terraform-google-modules/terraform-google-project-factory/releases/tag/v0.1.0) was published Jan 23, 2019. Both seem like high-quality modules and both are maintained and published by Google.

As it may be difficult to change later, we would like to make sure we're choosing the most suitable approach for maintaining our infrastructure. Should we prefer cloud-foundation-fabric, terraform-google-project-factory, or a mix of the two?

**Comment by juliocc** [comment]
Fabric provides leaner modules which are meant to be used as a starting point (i.e. users are expected to fork the repo and add their own customizations) whereas the rest of the terraform-google-modules provide more heavy-weight implementations covering more nuanced use cases.

Fabric modules are designed to be composed with other Fabric modules easily. You could mix them with other modules (or naked resources) but that's not something we regularly test for. If you want to go that route, you'll notice that most Fabric modules have an output with the resource as returned by the provider. This might come handy if you need to extract an attribute that is not exposed directly by any other output.

**Comment by drebes** [comment]
Adding to what Julio said, we don't document the rationale for fabric modules one by one, but we do it for the whole module set of fabric. From the README:
> The whole repository is meant to be cloned as a single unit, and then forked into separate owned repositories to seed production usage, or used as-is and periodically updated as a complete toolkit for prototyping. You can read more on this approach in our manifesto.

> Both the examples and modules require some measure of Terraform skills to be used effectively. If you are looking for a feature-rich black box to manage project or product creation with minimal specific skills, you might be better served by the Cloud Foundation Toolkit suite of modules.

Additionally, the [MANIFESTO](https://github.com/terraform-google-modules/cloud-foundation-fabric/blob/master/MANIFESTO.md) provides the general guidelines and values that these modules follow. If these values are important for your project, fabric modules are generally a nice choice.

**Comment by jawnsy** [comment]
Thanks to you both for the quick responses! That really helps me understand Fabric better. There's still some things I'm a bit unclear on, so to try to distill down my questions better:

1. How would you compare/contrast a module like [Project Factory](https://github.com/terraform-google-modules/terraform-google-project-factory) with the [Cloud Foundation Fabric Project](https://github.com/terraform-google-modules/cloud-foundation-fabric/tree/master/modules/project) module? I suppose the answer is that former is an opinionated approach, and the latter is a flexible/unopinionated one that one can compose very customizable resources with?
2. From the links in [Cloud Foundation Toolkit](https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit), it seems that Project Factory is a _part_ of that toolkit, is that true? That wasn't clear to me, particularly since I'd have expected anything in Cloud Foundation Toolkit to reuse modules from Cloud Foundation Fabric in that case, but it seems to use raw resources directly
3. For small teams that want to set things up according to Google's published best practices, are the Cloud Foundation Toolkit projects the "preferable" approach? Smaller teams do better with opinionated approaches, since they don't have specific needs and it keeps them on a well-tested path
4. Is it possible to migrate between Cloud Foundation Toolkit and the Cloud Foundation Fabric project, for example, if we start with Cloud Foundation Toolkit and later realize that we need more customizability?
5. For customers using Cloud Foundation Fabric directly, what are the recommendations for tracking new features and updates to modules? If they customize modules by forking and modifying the code as suggested in the README, will they need to review and rebase changes later?

**Comment by bharathkkb** [comment]
@juliocc Looks like we could add some docs on this since this is the second issue on the topic https://github.com/terraform-google-modules/cloud-foundation-fabric/issues/207
 
> Fabric modules are designed to be composed with other Fabric modules easily.

This is the case for CFT modules as well. Many of our [examples](https://github.com/terraform-google-modules/terraform-google-project-factory/blob/master/examples/shared_vpc/main.tf) and [example-foundation](https://github.com/terraform-google-modules/terraform-example-foundation) showcase this.

> How would you compare/contrast a module like Project Factory with the Cloud Foundation Fabric Project module? I suppose the answer is that former is an opinionated approach, and the latter is a flexible/unopinionated one that one can compose very customizable resources with?

I would look at what each module does and your use cases. Project Factory has docs around [features](https://github.com/terraform-google-modules/terraform-google-project-factory#features) along with examples.

> From the links in Cloud Foundation Toolkit, it seems that Project Factory is a part of that toolkit, is that true? That wasn't clear to me, particularly since I'd have expected anything in Cloud Foundation Toolkit to reuse modules from Cloud Foundation Fabric in that case, but it seems to use raw resources directly

Yes, Project Factory is part of Cloud Foundation Toolkit(CFT). Cloud Foundation Fabric has a [different approach](https://github.com/terraform-google-modules/cloud-foundation-fabric/blob/master/MANIFESTO.md) which makes reuse within other CFT modules complicated. We do use other CFT modules like Project Factory within [other modules](https://github.com/terraform-google-modules/terraform-google-bootstrap/blob/56dec3e4e0496b01231e8e37c60e43b17dddaff0/main.tf#L57).

> For small teams that want to set things up according to Google's published best practices, are the Cloud Foundation Toolkit projects the "preferable" approach? Smaller teams do better with opinionated approaches, since they don't have specific needs and it keeps them on a well-tested path

CFT modules are opinionated and are expected to be used directly. If its a smaller team this maybe more preferable than  having to create in house modules and test them. 

But we also have more advanced CFT users who fork and modify or vendor into other modules to make in house modules. We still suggest these users still try to keep up with upstream.  Since I am not a Fabric Maintainer, I'll defer to @juliocc 

> Is it possible to migrate between Cloud Foundation Toolkit and the Cloud Foundation Fabric project, for example, if we start with Cloud Foundation Toolkit and later realize that we need more customizability?

For the projects use case, I would recommend importing the project using `terraform import` and then letting the respective module create any additional resources.

For customers using Cloud Foundation Fabric directly, what are the recommendations for tracking new features and updates to modules? If they customize modules by forking and modifying the code as suggested in the README, will they need to review and rebase changes later?

defering to @juliocc

**Comment by juliocc** [comment]
> Thanks to you both for the quick responses! That really helps me understand Fabric better. There's still some things I'm a bit unclear on, so to try to distill down my questions better:
> 
> 1. How would you compare/contrast a module like [Project Factory](https://github.com/terraform-google-modules/terraform-google-project-factory) with the [Cloud Foundation Fabric Project](https://github.com/terraform-google-modules/cloud-foundation-fabric/tree/master/modules/project) module? I suppose the answer is that former is an opinionated approach, and the latter is a flexible/unopinionated one that one can compose very customizable resources with?

Both are opinionated but in different ways. Fabric's approach and core principles are reflected [here](https://github.com/terraform-google-modules/cloud-foundation-fabric/blob/master/MANIFESTO.md). 
One important difference is that Fabric modules are always released and tested together, to ensure they support the same version of Terraform and GCP provider. This way we can ensure any interface change is reflected in every module and example.

> 2. From the links in [Cloud Foundation Toolkit](https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit), it seems that Project Factory is a _part_ of that toolkit, is that true? That wasn't clear to me, particularly since I'd have expected anything in Cloud Foundation Toolkit to reuse modules from Cloud Foundation Fabric in that case, but it seems to use raw resources directly

Yes, Project Factory is part of CFT. CFT and Fabric are related but both projects have different design goals as mentioned above.

> 3. For small teams that want to set things up according to Google's published best practices, are the Cloud Foundation Toolkit projects the "preferable" approach? Smaller teams do better with opinionated approaches, since they don't have specific needs and it keeps them on a well-tested path

I don't think team size is the defining factor. Following Google's best practices is orthogonal to the set of Terraform modules you select. Depending on what you're trying to achieve, your design might be easier or harder to implement with Fabric compared to CFT. 

I suggest you try both and select the one that best fits your needs. 

> 4. Is it possible to migrate between Cloud Foundation Toolkit and the Cloud Foundation Fabric project, for example, if we start with Cloud Foundation Toolkit and later realize that we need more customizability?

Everything is possible ;) But realistically, migrating from one to the other is probably going to be somewhat painful. You'll probably have to play around with `terraform import` and tweak the arguments a little to make sure you don't destroy anything accidentally.

> 5. For customers using Cloud Foundation Fabric directly, what are the recommendations for tracking new features and updates to modules? If they customize modules by forking and modifying the code as suggested in the README, will they need to review and rebase changes later?

The idea behind Fabric's fork-and-own recommendation is precisely to allow us to evolve while not impacting users as we introduce new features. That being said, incorporating upstream changes to your local modules will probably involve some level of manual review as you're suggesting.

**Comment by juliocc** [comment]
Closing this. @jawnsy feel free to reopen if you have further questions

**Comment by jawnsy** [comment]
Thanks all for the replies! Please let me know if you'd like me to open a pull request with some doc changes to include the above :)
